### PR TITLE
libc: add strtok implementation

### PR DESCRIPTION
**Pull request by sidcha**
This is a standard function and useful for applications.

Signed-off-by: Siddharth Chandrasekaran <siddharth@embedjournal.com>

### CODE REVIEW

**Review by pabigot — CHANGES_REQUESTED** [review]
Needs tests.

**Comment by cbsiddharth** [comment]
@pabigot, added tests.

File: lib/libc/minimal/source/string/string.c [hunk 1]
```diff
@@ -170,6 +170,37 @@ int strncmp(const char *s1, const char *s2, size_t n)
...
+ * sections. Consecutive `sep` chars in `str` are treated as a single
+ * separator.
+ *
+ * @return pointer to NULL terminated string or NULL on errors.
+ */
+char *strsep(char **str, const char *sep)
+{
+	char *start, *end;
+
+	if (str == NULL || *str == NULL || *str[0] == '\0') {
```
Comment (pabigot) [inline lib/libc/minimal/source/string/string.c:184]: `strsep` on Linux segfaults if `str` is a null pointer: that is an API mis-use.  It should probably do the same here.

Also it does not return null if `*str[0]` is zero: it returns the original value of `*str` and sets `*str` to NULL.

File: tests/lib/c_lib/src/main.c [hunk 2]
```diff
@@ -558,6 +558,45 @@ void test_tolower_toupper(void)
 	zassert_equal(strcmp(lw, tolw), 0, "tolower error");
 }
 
+void test_strsep_do(char *str, char *sep, int tlen, char **toks, bool expect)
```
Comment (pabigot) [inline tests/lib/c_lib/src/main.c:561]: This should be `..., const char **toks, ...`

File: lib/libc/minimal/source/string/string.c [hunk 3]
```diff
@@ -170,6 +170,37 @@ int strncmp(const char *s1, const char *s2, size_t n)
...
+		return NULL;
+	}
+
+	start = *str;
+	while (*start && strchr(sep, *start)) {
+		start++;
+	}
+	end = start;
+	while (*end && !strchr(sep, *end)) {
```
Comment (pabigot) [inline lib/libc/minimal/source/string/string.c:193]: This appears to be skipping over a sequence of separators without generating empty tokens, which is inconsistent with standard `strsep`.  See comment in test case.

File: tests/lib/c_lib/src/main.c [hunk 4]
```diff
@@ -558,6 +558,45 @@ void test_tolower_toupper(void)
...
+	const char **tc01 = { "1", "2", "3", "4", "5" };
+
+	test_strsep_do("1,2,3,4,5",           ",",  5, tc01, true);
+	test_strsep_do(",, 1 ,2  ,3   4,5  ", ", ", 5, tc01, true);
+	test_strsep_do("1,,,2 3,,,4 5",       ", ", 5, tc01, true);
+	test_strsep_do("1,2 3,,,4 5  ",       ", ", 5, tc01, true);
+	test_strsep_do("0,1,,,2 3,,,4 5",     ", ", 5, tc01, false);
+	test_strsep_do("1,,,2 3,,,4 5",       ",",  5, tc01, false);
+	test_strsep_do("A,,,2,3,,,4 5",       ",",  5, tc01, false);
+	test_strsep_do("1,,,2,3,,,",          ",",  5, tc01, false);
```
Comment (pabigot) [inline tests/lib/c_lib/src/main.c:596]: With real `strsep` this produces 11 tokens (parenthesized value is the value of the pointer referenced by the first argument)
```
tok 1: '1' (0x7ffc1d0184bf)
tok 2: '' ((nil))
tok 3: '' (0x7ffc1d0184c0)
tok 4: '' (0x7ffc1d0184c1)
tok 5: '2' (0x7ffc1d0184c3)
tok 6: '' ((nil))
tok 7: '3' (0x7ffc1d0184c5)
tok 8: '' ((nil))
tok 9: '' (0x7ffc1d0184c6)
tok 10: '' (0x7ffc1d0184c7)
tok 11: '' ((nil))
```

File: tests/lib/c_lib/src/main.c [hunk 5]
```diff
@@ -558,6 +558,45 @@ void test_tolower_toupper(void)
...
+			      "strsep error '%s' / '%s'", str, sep);
+	} else {
+		zassert_not_equal(len, tlen,
+				  "strsep error '%s' / '%s'", str, sep);
+	}
+}
+
+void test_strsep(void)
+{
+	const char **tc01 = { "1", "2", "3", "4", "5" };
```
Comment (pabigot) [inline tests/lib/c_lib/src/main.c:587]: You can't initialize a pointer this way.  It needs to be:

    const char *tc01[] = { "1", "2", "3", "4", "5" };

**Review by pabigot — CHANGES_REQUESTED** [review]
The test produces build errors and segfaults on native_posix on Ubuntu, where the host C library is being used, so apparently the test isn't valid.  Also the implementation is not consistent with standard `strsep`.

Finally turns out `strsep` is not a standard C function, it's a BSD extension.  From the Linux man page:
```
NOTES
       The  strsep()  function was introduced as a replacement for strtok(3), since the latter
       cannot handle empty fields.  However, strtok(3) conforms to C89/C99 and hence  is  more
       portable.
```

I don't think we should add this to Zephyr.    We could add `strtok`, though.

File: tests/lib/c_lib/src/main.c (hunk 5 above)
Comment (cbsiddharth) [reply tests/lib/c_lib/src/main.c:587]: Sorry. That was a sleepy check checkpatch fixup. Will correct it.

File: tests/lib/c_lib/src/main.c (hunk 4 above)
Comment (cbsiddharth) [reply tests/lib/c_lib/src/main.c:596]: I will change this to strtok_r so the behavior remains same.

File: lib/libc/minimal/source/string/string.c (hunk 1 above)
Comment (cbsiddharth) [reply lib/libc/minimal/source/string/string.c:184]: I'll make sure new strok_t implementation complies with the specs. Thanks for the review.

Comment (pabigot) [reply lib/libc/minimal/source/string/string.c:184]: You're welcome.  I do think `strtok_r` will be a good enhancement, and look forward to seeing that available.

File: lib/libc/minimal/include/string.h [hunk 6]
```diff
@@ -25,6 +25,7 @@ extern size_t strlen(const char *s);
 extern size_t strnlen(const char *s, size_t maxlen);
 extern int    strcmp(const char *s1, const char *s2);
 extern int    strncmp(const char *s1, const char *s2, size_t n);
+extern char  *strtok_r(char *str, const char *sep, char **sotre);
```
Comment (pabigot) [inline lib/libc/minimal/include/string.h:28]: s/sotre/store/ (or follow the [POSIX declaration](https://pubs.opengroup.org/onlinepubs/9699919799/functions/strtok.html) and make it `state`)

**Review by pabigot — APPROVED** [review]
Looks good, thanks.

A case could be made to also provide `strtok` since it's trivial to do.  I won't insist but would support it if others think it worthwhile.

File: lib/libc/minimal/include/string.h (hunk 6 above)
Comment (cbsiddharth) [reply lib/libc/minimal/include/string.h]: Changed all store refs to `state`.

**Comment by cbsiddharth** [comment]
> A case could be made to also provide `strtok` since it's trivial to do. I won't insist but would support it if others think it worthwhile.

Adding support is trivial but strtok advocates bad programming practice IMO (I for one, always use `strtok_r`). But do let me know if it should be added for legacy/completeness sake.

**Comment by andrewboie** [comment]
> > A case could be made to also provide `strtok` since it's trivial to do. I won't insist but would support it if others think it worthwhile.
> 
> Adding support is trivial but strtok advocates bad programming practice IMO (I for one, always use `strtok_r`). But do let me know if it should be added for legacy/completeness sake.

There's no requirement, we add stuff to minimal libc more or less on an as-needed basis. We can and should omit bad stuff like `strtok()` and `gets()`. 

One of these days we should look into pulling in an OS-agnostic test suite for C library functions to exercise minimal libc testing rather than roll our own.

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**